Line-wise author info (resolved #77)

GollyTicker opened the pull request:
This is a **work in progress PR!**

@Vinzent03 Could you take a look at the first two commits only?
For working with the plugin, I've added a dockerized environment to reproducibly setup the node packages. I also added a [test-vault](https://github.com/GollyTicker/obsidian-git-test-vault-online) as a submodule. 
If you think, they are useful, I can extract them into separate branches and have them merged into the plugin for reproducability and makiing it easier for others to contribute. Otherwise, I can simply remove them from the branch and keep them just to me for development purposes.
What do you think?

Regarding the actual implementation, I will ping in actual issue, when I have a relevant working state to disucss.

Vinzent03 commented:
Hey thanks for working in this feature, but I really don't get why you are using docker. May you explain this further?

GollyTicker commented:
I am using docker for a reproducible development environment. I use this in most larger projects everywhere, because then I don't need to deal with any OS-level global installations which can create a dependency hell. If you don't like it, I can also remove it from the branch and keep it private only. However, for new developers it's always useful to have such dev-infrastructure. I can add more explanations and a README for that, if necessary.

Furthermore, I've a working basic prototype for the line authoring view. This is how it looks like right now:
![demo-1](https://user-images.githubusercontent.com/4527862/187084588-8ee0044f-ce66-4016-bbbc-ee655f2afa63.gif)

Again. This is totally not finished. The UI can be improved. And the delay between chaning the lines in the editor and the line author view being updated is because Obsidian saves the changes to the filesystem only about a second after the last keystroke. So the git blame is delayed by that amount. I have ideas to work around that, but will implement them when the time is right.

Would you like to take an early look into the code and discuss any things there? the code is not finished, but most of the moving parts are set - i.e. the basic architecture and structure. Documentaiton, extensive testing and co is still remaining though. And all the variations such as the colorbar without the author etc. and copying the revision are also to be done.

What do you think?

Taitava commented:
@GollyTicker Your GIF looks nice to me! 🙂 How does it combine with the linenumbers column if you turn that on in the editor settings (_Show line number_)?

Perhaps I'd suggest to make the author information to be visually separated from the content, i.e. by making the column to have a different background color (maybe the same as the linenumber column). Anyway, this is just tweaking, the main thing is that the actual information is there and it works really well. Good job! 🌞

GollyTicker commented:
@Taitava Currently it's simply displayed side by side to the line number. I'm planning to improve it in the way, that the condenced version (pure color withouth hash, date and author) should simply color the line numbers, if they exists, or just a blank area if they don't exist.

Regarding the UI of the line authoring. I haven't optimised the UI at all yet. Just wanted to show, that I've completed most fo the "hard work" of parsing the git-blame and integrating it into the view. I'll try out different styles and post a few of them when the time is right to get further feedback.

GollyTicker commented:
**Update: Date and time is now configurable**

# Date defaults
![date-defaults](https://user-images.githubusercontent.com/4527862/189494924-0a3d214d-68c8-4cb1-842a-71c8f12740c0.png)
and its settings
![date-defaults-settings](https://user-images.githubusercontent.com/4527862/189494931-2f675c4f-1ebb-429f-b69e-94e1e26900d0.png)

 # Date and time
Uses local timezone by default
![date-time](https://user-images.githubusercontent.com/4527862/189494937-1dbcdfc2-6458-4516-a1ff-e14463ef72c8.png)
and its settings
![date-time-settings](https://user-images.githubusercontent.com/4527862/189494951-a8b8b367-1511-46e2-ae52-dc67ed80fede.png)

# Date and time with explicit UTC timezone
![date-time-utc](https://user-images.githubusercontent.com/4527862/189494957-254c1051-1ee3-402b-b718-83b496bc808f.png)
and its settings
![date-time-utc-settings](https://user-images.githubusercontent.com/4527862/189494964-73c40967-e52f-43fa-b52f-7f5120fb35d0.png)

# Settings 
The settings are only shown, when the line author info is active. There are a couple of options for everything. The date formatting has a live preview as well.
![demo-2-date-settings](https://user-images.githubusercontent.com/4527862/189494992-118a3f8b-32b4-4354-b425-baa1c7d274c8.gif)

GollyTicker commented:
**Update: Improved visuals and colorbar only option**
New improved styling
![colorbar-visuals](https://user-images.githubusercontent.com/4527862/189522854-0074c8a9-3e65-41bf-bbcf-e6df3799ec66.png)

Only the colorbar
![colorbar-only-visuals](https://user-images.githubusercontent.com/4527862/189522858-ed8ab74e-4f74-4fda-8fc8-a0667f7ffe2f.png)

Also works in submodules now!
![submodule](https://user-images.githubusercontent.com/4527862/189522897-931b28ef-b2e4-4038-bf73-9130ef4a1f65.png)

Taitava commented:
The screenshots look really nice! 🌞

GollyTicker commented:
@Vinzent03 I've implemented all must-features and would like to go through the code with you once. I'm currently writing the documentation and architecture to improve the understand-ability of the code. However, perhaps it makes sense, that we also talk before I write the complete documentation - so that we can discuss some architectural from your perspective. I also have a few questions on a few small occasions where I'm not sure what's the correct obsidian-plugin-like way to implement things. Finally, there are no tests yet and I want to add tests for at least the most complex logic stuff.

What do you think?

If you want to play around with the current status, just run `scripts/restart-dev.sh`, check that the build watcher succeded and open `test-vault` as a vault in Obsidian.

Vinzent03 commented:
Sure that's a great idea. Do you have a Discord account so we can discuss this further?

GollyTicker commented:
## Dark and Light mode
![dark-light-mode](https://user-images.githubusercontent.com/4527862/191110076-9394a166-8898-415d-86af-26bc9bec277c.gif)

## Copy Commit Hash
![clipboard](https://user-images.githubusercontent.com/4527862/191110376-c97c2c4e-1fb4-4b02-b1b4-142d2f843783.gif)


## Adaptive Color and Unintrusive Text Filler until result is ready
**Before**
![adaptive-color-and-text-before](https://user-images.githubusercontent.com/4527862/191112176-3346b17d-4cfc-480e-9d87-fc8e9362c22b.gif)


**After**
![adaptive-color-and-text-after](https://user-images.githubusercontent.com/4527862/191111132-71ac2010-2743-4b8f-b67b-ff40ed22ca9f.gif)

It was quite some effort to make this adaptive coloring work. It was important to me, that the update of the UI when the result is ready is as unintrusive as possible. Opening a new file is done very often, so the UI update should be not distracting.

Taitava commented:
Nice looking, again! 👍

GollyTicker commented:
@Vinzent03 I'm finally ready for the feature review. :partying_face: :partying_face: :partying_face:  Would you like to take a look at all of the changes? We can have a call sometime again on the general procedure.

I'd next ask of beta-testers in the original issues. And when the feature is working robustly, I we can then integrate it into master.

Before that, however, I would also like to finish the [feature set docs for the public](https://github.com/GollyTicker/obsidian-git/blob/line-author-info/docs/public/LineAuthorFeature.md) containing quick examples and GIFs.

Before the merge, I'd also rebase all into just a few commits.

GollyTicker commented:
## Be a **beta tester** for this feature!

We are happy, if a few people are willing to install this version of the plugin and give some early feedback.
I want to make sure, that the feature is working reliably before we publish it.

## How to install
1. Make sure, that you are currently using the most updated `obsidian-git` plugin.
2. Make a commit with your current state and remember it's revision in case you want to return there.
3. Close Obsidian
4. Open the folder `.obsidian/plugins/obsidian-git` and replace the existing `main.js` and `styles.css` with the files from [v1.zip](https://github.com/denolehov/obsidian-git/files/9641467/v1.zip)
5. Re-open Obsidian
6. Activate the Line Author Feature in `Settings > Obsidian Git > Line author information`.
7. Play around with everything :tada: 

If you want to revert your installation - you can simply replace `main.js` and `styles.css` with the files from step 2. Note, that updating your obsidian-git will overwrite these files and the line author feature will not be there then.

## How to use
The usage of the plugin should be mostly self-explanatory. A quick summary of example usages is being written now here:
https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature

You are encouraged to simply use it in your normal workflows. If you find something, then *please* make a [new issue here](https://github.com/denolehov/obsidian-git/issues/new/choose) or simply *comment* in this pull request.

My priority is to have the most relevant bugs fixed before publish.

You can also write enhancements, but they'll be implemented after the merge (unless they are very important).

Allright then! Looking forward to your feedback! :partying_face: :muscle:

Vinzent03 commented on tsconfig.json:
```diff
@@ -11,11 +11,16 @@
             "ES7"
+        ],
+        "typeRoots": [
+          "src",
```
Why did you add `src`. I have no problems finding the types in `externalLibTypes.d.ts` without this setting.

GollyTicker replied:
Ah, I didn't know that it would work without it as well and I didn't test it.

I'll try it out and remove it if it works.

GollyTicker commented:
:information_source: I've done a few small fixes and minor improvements in [v2.zip](https://github.com/denolehov/obsidian-git/files/9649762/v2.zip). Please use this zip from now on.

If you already had `v1` running locally, then (after installing `v2`) you need to go to the Obsidian-Git Line Author Information settings and set the time-zone config explicitly again as it's values have changed.

Also, the feature guide for users is done: https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature

@Vinzent03 How could we merge the feature guide into the main repository?

Vinzent03 reviewed (commented):
Does the context menu work for you? `registerLastClickedGutterHandler` is called, but the event doesn't get fired for me.

Can you please remove the submodule, because it's not related to the plugin dev itself.

GollyTicker commented:
@Vinzent03 The `registerLastClickedGutterHandler` works perfectly fine with me. I had it installed by a few friends and we havent had any issues until now. However, I didn't ask them to specifically test this. I'm on Linux/Ubuntu and Obsidian v0.15.9. But I didn't text it myself on any other OS or version.

Regarding the submodule. I suggest you to consider actually leaving the test-vault inside.
[The dataview plugin also does this, for instance.](https://github.com/blacksmithgu/obsidian-dataview/tree/master/test-vault) (Though it's simply a directory there and not a submodule.)

I used the text-vault a lot for testing purposes, and I think that it can simplify all kinds of dev processes. The images for the [user guide](https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature) were all generated with the test-vault. Making it a submodule allows one to not have it forcefully cloned, when people are only interested in the code. However, making it a submodule allows one to synchronize the specific stage of development with it's submodule. After all, the test-vault is also part of the development process, so it makes sense to include it.

I'd also like to have a call with you on the general next steps before the final merge and release. We can talk about that then as well.

GollyTicker commented:
* [ ] Somtimes, it takes longer to compute the line authoring. During that time the `%` placeholders as shown. It is however possible to R-click and copy the commit hash - which shouldn't be. It returns `dummy-commit`.

GollyTicker commented:
### [v3.zip](https://github.com/denolehov/obsidian-git/files/9742443/v3.zip) is out with minor improvements

GollyTicker commented:
I could resolve the problem mentioned in https://github.com/denolehov/obsidian-git/pull/288#issuecomment-1270286575.
The performance is not better when actively working with the document.

GollyTicker commented:
Obsidian v1.0 was just released. The many UI changes however resulted in a few things that don't work anymore. I've added todos in the code where these problems needs to be resolved. Probably this feature should be right release with the next update of the plugin to the v1.0 release.

GollyTicker commented:
### [v4.zip](https://github.com/denolehov/obsidian-git/files/9781349/v4.zip) was added with Obsidian v1.0 support

It also contains some performance improvements.

@Vinzent03 I had the same bug as you had with the context menu on v1.0. I had to use a hack to resolve it. An obsidian team member confirmed the bug on discord and it'll be released in the next release.

GollyTicker commented:
I suspect a bug with Obsidian v1.0 which leads to the UI being messed up and obscuring parts of the editor text when line numbers are not also shown. [See the corresponding bug report](https://forum.obsidian.md/t/added-editor-gutter-overlaps-and-obscures-editor-content/45217) I'll attempt a quick-fix for that for now.

MateusMolina commented:
looking forward to it. NIce work so far!

GollyTicker commented:
### :information_source: [v5.zip](https://github.com/denolehov/obsidian-git/files/9795451/v5.zip) with a UI fix

Taitava commented:
# Beta testing v5
## 1. My name does not appear completely.
_Author name display_ setting:
 - `First name`: Shows correctly my first name, _Jarkko_.
 - `Initials`: Shows _J_, but should show _JL_.
 - `Last name`: Shows my first name, not last name.
 - `Full name`: Shows my first name only. Example screenshot below:
    ![kuva](https://user-images.githubusercontent.com/13002142/197010595-6249e4c8-8ad1-4545-ab67-7b89cfc03b5d.png)

Here you can see in terminal that I have configured my name correctly:
![kuva](https://user-images.githubusercontent.com/13002142/197009997-c5b7e148-d392-48a7-9a53-560f30957fe9.png)

--------------

I'll edit this post if I'll come up with more comments today. I just started testing, looks promising so far! 🌞

GollyTicker commented:
@Taitava Thanks a lot for the detailed beta feedback! I highly appriciate that

## [v6.zip](https://github.com/denolehov/obsidian-git/files/9834932/v6.zip) with fixes of [1&2 named here](https://github.com/denolehov/obsidian-git/pull/288#issuecomment-1285871285)

Concretely, I fixed the full and last name bug - and also added a new option to ignore whitespace and newlines in the settings. Since, people might actually be interested in indentation changes, I decided to not ignore whitespace by default. (Also, this is the default by git anyways.)

@Taitava Regarding 3. This will take a quite more effort, as we would need to add corresponding logic in the CodeMirror section ans continously update the gutter view. I by delaying the gutter view update, I could strongly reduce the implementation complexity at the cost, that the UI is shortly inconsistent. A good point you mentioned. Depending on how long the review takes, I might implement this before the merge - or later afterwards.

>  Nice to see someone spending time to deliver good quality. :sunglasses: :+1: 

Thanks :blush:

MateusMolina commented:
Hi,

I've been testing it since last week and found no bugs besides the ones mentioned by @Taitava. This feature is really helpful to check whether a file is old or not and how much work you put into it.

From my side, I miss a way to 'collapse' the bar in a way that only a color strip is shown. In my opinion, this is as helpful and less intrusive as the current implementation - especially, when you have many files opened at the same time. Additionally, when the user hovers the cursor on the strip it expands revealing the date and committer.

Another helpful feature would be to be able to see the git diff when clicking a particular date. 

All in all, I really appreciate the effort you put into this feature! Amazing job! :)

GollyTicker commented:
# [v7.zip](https://github.com/denolehov/obsidian-git/files/9847391/v7.zip) with quick update of gutter while editing

@Taitava I think displaying `%` is a good idea. So I went ahead and implemented it. This is how it looks like :wink: 

![](https://github.com/GollyTicker/obsidian-git/wiki/assets/line-author-soft-unintrusive-ux-editing.gif)

It's also explained in the [wiki](https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature#soft-and-unintrusive-ansynchronous-view-updates) (not merged yet).

@MateusMolina  Thanks for the feedback and your responses :smiley: 
Regarding the colorbar. That part is aleady possible. [You can right-click and hide the hash/author/date to only show the colorbar.](https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature#quick-configure-gutter)

Your other suggestions are already in this [list of potential future improvements](https://github.com/GollyTicker/obsidian-git/blob/line-author-info/docs/dev/LineAuthorFeature.md#potential-future-improvements). I've not yet decided when and how to tackle them. Ideally, I'd have the feature out and published first before diving into them.

GollyTicker commented:
# [v8.zip](https://github.com/denolehov/obsidian-git/files/9873696/v8.zip) with quick gutter update fixes
* when actively typing in larger vaults (like mine), sometimes the gutters were messed up. fixed.
* when wroking with empty files, the quick gutter update was also leading to errors.

However, I also found a new bug which I haven't yet been able to fix:
* [ ] when opening the vault directly inside an empty and tracked file, the gutter spacing during typing is wrong and leads to distracting ui updates.

Taitava commented:
Thank you @GollyTicker for continuing to make improvements! ❤️ Sorry that I've away for a while.

# Collapse the bar

@MateusMolina wrote:
> From my side, I miss a way to 'collapse' the bar in a way that only a color strip is shown.

As @GollyTicker answered, it's possible already, but I'd like to fine-tune it a bit: When right cliking the line number column, I'd add a menu option like _Show line commit details_ (or something) and if it's checked, then show the three options:
 - _Show commit hash_
 - _Show author name_
 - _Show date_

If it's not checked, hide the those options and shrink the column so it's narrow.

This would be because I like to show both author name and date, but sometimes I want to save space and hide the commit details. I might switch between showing and hiding commit details quite often, so it would be easier if I only need to use one switch instead of two for hiding or showing the two pieces of information.

# Previous comment's things

Regarding my [previous comment](https://github.com/denolehov/obsidian-git/pull/288#issuecomment-1285871285), now that I'm testing `v8`:
1. I can confirm that the name displaying is fixed, works really well! 🙂 
2. The `+++` showing on an old line when adding a line to bottom of the file is a bit problematic (even if _Ignore whitespace and newlines in changes_ is off intentionally), because it indicates that the line would be _completely_ new, even though it's actually _changed_ (the change is that a newline character is added to the end of the line, I realized now). Turning the ignore whitespace setting on does help, but would it be hard to make it so that when whitespace is not ignored, show the last line as changed instead of new when another line is added after the last line? Sorry if I wrote this in a messy way.
3. Thank you, the `%%%` thing is working perfectly now. 🙂👍 

# Text color

I'm using [Red graphite theme](https://github.com/seanwcom/Red-Graphite-for-Obsidian) with base color theme being set to _Light_. (I used a different, dark theme in my previous comment, so this thing did not come to my mind then). The text color is not so easy to read with this theme:

![kuva](https://user-images.githubusercontent.com/13002142/201537656-9cca5cb1-ae61-4690-b24e-6c7b3846d1bf.png)

How to solve:
- First off, I can try to put some dark background colors for old and new commits in the settings, but it's not so nice as it limits my choice of colors.
- Add a text color field to the settings. The downside is that it creates yet another setting, and it's still a single color, that can show well on one background, but badly on another.
- Calculate the text color automatically. [This SO answer describes some theory how to do it](https://stackoverflow.com/a/3118280/2754026), but at least I felt a bit overwhelmed when reading it quickly. Perhaps just use inverted background color (255 - background R, and same for G and B), but switch to either black or white if the resulting text color would be too near the background color (happens for middle range colors, e.g. rgb(127,127,127)).
- You have added a CSS class `obs-git-blame-gutter`. I could create a CSS snippet that would change that class's `color` property to whatever I want. But it still has the issue that I could only pick one color regardless of the background. At least I'm not aware of if it's possible to define a CSS color that would reflect to the background.

GollyTicker commented:
Responding to @Taitava:

> Thank you @GollyTicker for continuing to make improvements! heart Sorry that I've away for a while.

Thanks for your feedback. I was busy as well and will also be somewhat busy in the near futuer - but I'll try to look into all ofthis by end of year and then beginning of next year.

---

Regarding **Collapse the bar**
> This would be because I like to show both author name and date, but sometimes I want to save space and hide the commit details. I might switch between showing and hiding commit details quite often, so it would be easier if I only need to use one switch instead of two for hiding or showing the two pieces of information.

You can actually also use the Command Palette `Obsidian Git: Toogle Line Author Information` to toggle the view.
See: https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature#activate

I see your use case. However, I am not sure, what kind of UI would be best for the largest set of users. In that sense, I prefer to ship this bar the way it is now - and to provide further improvements when more feedback is there from the users.

---

**Previous comment's things**
Happy to hear, that you appriciate the updates and changes.

Regarding
> ... but would it be hard to make it so that when whitespace is not ignored, show the last line as changed instead of new when another line is added after the last line?
In theory, making such a hack wouldn't be quite possible. However, my intention was to keep the obsidian functionality as close to what on would experience when using git via scripts. Essentially, there are two conflicting goals. First, we want to stay close to how git works and avoid showing different diffs than what git would show natively (ensuring cross-compatibility). Second, we want to improve the UX and remove distracting technical details which don't matter to the user (e.g. a new 'newline' at the end of a line shouldn't count as a change).

Right now, I'd rather have the feature shipped into a release and think about adding an option for this, which switched between compatability-mode vs. comfort-mode. What do you think on that?

* [ ] Add compatibility-mode vs comfort-mode as setting for changes display as potential futture feature

---

Regarding **Text color**: Good point. Thanks for presenting a few solution approaches as well as the idea with automatic calculation of the color. I am thinking of an approach, where each gutter would define a CSS variable `--obs-git-age` with values between `0.0` and `1.0`. Then one could define the text-color as a function of these variables.

Right now, the background color is simply [a linear interpolation conditioned on light/dark mode](https://github.com/GollyTicker/obsidian-git/blob/line-author-info/src/lineAuthor/view/gutter/coloring.ts#L54). I could also simply use a linear formula for the text color.

While the CSS variable approach is more flexible, it's also harder to design and ensure quality.

I prefer sticking with the simpler manual linear interpolation first - and then see how the feedback from the community is after the featureis realeased.

* [ ] fix text-coloring

---

@Vinzent03 Thanks for merging the latest master.

GollyTicker commented:
@Taitava I added the text color setting. The field is simply used as the CSS color - and hence it can implicitly also use CSS variables. I realised, that I was using `var(--text-muted)` - which works well for the default Obsidian themes. But for your theme `var(--text-on-accent)` works better. So now you can set the text color css to this value and then it'll look like this:

![Screenshot from 2022-12-22 01-44-19](https://user-images.githubusercontent.com/4527862/209030741-d5d939c0-bdc9-4792-b385-2426348bf923.png)

Even, if I hadn't made this change, you could have also done the change to the text and the background (!) yourself by adding this custom CSS:
```css
.obs-git-blame-gutter {
    background-color: var(--background-primary-alt) !important;
    border-color: var(--background-primary-alt) !important;
}

.obs-git-blame-gutter > div {
    color: var(--text-muted) !important;
}

```
It overrides [whatever is specified here](https://github.com/GollyTicker/obsidian-git/blob/line-author-info/styles.css#L488). I would rather focus on optimising the UI for the default themes, because integrating a good UI with more themes can get really complex and difficult quickly.

I don't know how else to make theme-integration possible, without it getting impractical.

Taitava commented:
> You can actually also use the Command Palette `Obsidian Git: Toogle Line Author Information` to toggle the view.
> See: https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature#activate

Thanks! This is good 🙂 

> Right now, I'd rather have the feature shipped into a release and think about adding an option for this, which switched between compatability-mode vs. comfort-mode. What do you think on that?

It's a good idea to get this completed with minimum requirements. There's already a good amount of stuff implemented. More improvements can be done later, if needed, like you said. 👍 

> I added the text color setting. The field is simply used as the CSS color - and hence it can implicitly also use CSS variables.

Thank you! 😎

GollyTicker commented:
# [v9.zip](https://github.com/denolehov/obsidian-git/files/10297457/v9.zip) with minor fixes
* text color css is now adaptable in the settings
* empty tracked file behaves correctly when opened as the first file on startup

Taitava commented:
Nice! 🌞 I like the text color setting! 🙂

GollyTicker commented:
@Vinzent03 When do you think could we have the final steps onto merging this?

Vinzent03 commented on src/lineAuthor/lineAuthorProvider.ts:
```diff
@@ -0,0 +1,100 @@
+     * 
+     * See {@link LineAuthoringId}
+     */
+    private lineAuthorings: Map<LineAuthoringId, LineAuthoring> = new Map();
```
From what I see `lineAuthorings` contains for every existed file state the blame. I think this can stack up in a longer Obsidian session. We don't need to store every blame of every ever opened file, right?

Vinzent03 replied:
Okay let's see if someone encounters issues.

tamiroh commented:
This is what I have wanted for a long time and I am really looking forward to its release! 👍

Vinzent03 commented on src/lineAuthor/view/gutter/commitChoice.ts:
```diff
@@ -0,0 +1,38 @@
+    return diff > 0;
+}
+
+function getAbsoluteAuthoringMoment(commit: BlameCommit): Moment {
```
I see I haven't described enough. From what I see it's only used in `isNewerThan()` above, where `commit.author!.epochSeconds` should be sufficient.

Vinzent03 replied:
https://github.com/denolehov/obsidian-git/pull/288/commits/37f98ded4325364bed7487261ec03a66e5a1d1bd That's what I meant.

GollyTicker commented on src/lineAuthor/view/gutter/commitChoice.ts:
```diff
@@ -30,14 +29,7 @@ export function chooseNewestCommit(
 }
 
 function isNewerThan(left: BlameCommit, right: BlameCommit): boolean {
```
Are you sure, that we need to check the epochSeconds and not also consider the timezone?

It's been a long time, since I last wrote this. But I am certain, that I've tested it back then and I think I would have tested it here to get the correct results. I am not sure, whether the new code and the old code are equivalent.

Vinzent03 replied:
According to your documentation of `epochSeconds`, it's in UTC. So comparing these should be sufficient. I even think the previous one was incorrect, because you want to compare the actual timestamp and not depend on timezone.

GollyTicker replied:
Okay, then maybe I was wrong back then. Thx for fixing!

GollyTicker reviewed (commented):
We should at least mention the test-vault somewhere in the dev-docs, as it's much easier to use that than to re-create testcases for the dates yourself. It was cumbersome sometimes to create some of these testcases.

GollyTicker commented on src/main.ts:
```diff
@@ -113,6 +113,8 @@ export default class ObsidianGit extends Plugin {
     }
 
     async refreshUpdatedHead() {
+        console.log("test");
+
```
Remove this test code.

Vinzent03 reviewed (approved):
I think this is ready now. @GollyTicker What do you think?

GollyTicker commented:
Looks good from my side!

Only a few small last things:

> * [ ] publish feature guide for users (merge wiki)
Here is the wiki: https://github.com/GollyTicker/obsidian-git/wiki/Line-Author-Feature
Can you also merge it?
> * [ ] decide whether to notify on the feature in the plugin
What do you think?

> * [ ] Add @GollyTicker as contributor for this feature in README
Would you like to add a contributor hint for that in the `Contact` section?

Vinzent03 commented:
The wiki is actually deprecated or outdated. The documentation now lies in `/doc` is published via Obsidian publish. Gonna add that.

I don't plan to add an in plugin notification, because it's a really advanced feature and I'm not really a fan of it. Though I will(or you if you want to)  add a Discord update message.

Sounds great adding you to the readme!

GollyTicker commented:
> The wiki is actually deprecated or outdated. The documentation now lies in /doc is published via Obsidian publish. Gonna add that.

Thank you. I refactored it a bit, to ensure it's properly integrated and linked.

> I don't plan to add an in plugin notification, because it's a really advanced feature and I'm not really a fan of it. Though I will(or you if you want to) add a Discord update message.

Yeah, I can understand that. If you can make a discord update message and link to the newest obsidian git version + link to the wiki page on line-authoring - that would be great!

> Sounds great adding you to the readme!

Thanks!

Looking forward to release this into the wild! 😊

Vinzent03 commented:
:tada: :tada:

Taitava commented:
Woohoo! 😀👍👍🎉🎉🎉

GollyTicker commented:
Released in 2.20.0 👉 https://github.com/denolehov/obsidian-git/releases/tag/2.20.0 🥳✨🥳✨

Update your local Obsidian-Git Plugin!

AmauryVanEspen commented:
Hello @GollyTicker 
really great feature !
could you tell us a bit more about interaction with submodule ? I'm a bit lost here with synchronization of the sub repository.
Cheers

GollyTicker commented:
@AmauryVanEspen 
Thanks for the appriciation of the feature!

Could you clarify, what kind of problems you have with submodules? [As mentioned in the documentation, submodules are supported natively and work correctly.](https://publish.obsidian.md/git-doc/Line+Authoring#Submodules+support).

I'm also privately using submodules and the files in the submodules show the correct line authoring information just like with top level files. Is there a specific workflow where you're hitting problems? Maybe you can show an example.

AmauryVanEspen commented:
Hello @GollyTicker 
it is maybe an issue with the set up of submodule. (see #438 )
the purpose is : 
I've a Vault,
A colleague has his own vault (with Obsidian Sync)
In my Vault, I've a folder "remote-blog" and an other folder "Sync With Git"
the remote-blog has it repo, sync with obsidian-git and github action to publish the dedicated content.
it works fine.
the second folder in the Vault is a submodule that i would like to use to sync some content with my colleague.
Line Authoring works locally.
However, the submodule does not sync through git properly. do not understand why.

GollyTicker commented:
@AmauryVanEspen .

Thanks for the explanation. It rather seems that the issue is related to the Git Submodules setup, as a working submodule is a pre-requisite, for Line Authoring working in that submodule.

For better debuggability of this issue and more importantly, that submodule issue, could you maybe create a small vault (with submodules) - which reproduces your problems? Please check, that cloning your bug reproducing repository on an independent computer (with your specific Obsidian+Git) version re-creates the problem. If you cannot reproduce your setup on an independent setup, then it might be an issue with your local configuration.

Here is my test vault I used for the Line authoring: https://github.com/GollyTicker/obsidian-git-test-vault-online
It also has submodules. Please clone the repository, including submodules via `git clone --recursive`, and open the vault in Obsidian. The Line Authoring should work in Obsidian correctly in the submodules. Let me know, if that works.
